weave/trace_server/clickhouse_trace_server_batched.py

```diff
@@ -292,10 +297,104 @@ def calls_query_stream(
         )
 
         select_columns = [c.field for c in cq.select_fields]
-        for row in raw_res:
-            yield tsi.CallSchema.model_validate(
-                _ch_call_dict_to_call_schema_dict(dict(zip(select_columns, row)))
-            )
+
+        if not req.expand_columns:
+            for row in raw_res:
+                yield tsi.CallSchema.model_validate(
+                    _ch_call_dict_to_call_schema_dict(dict(zip(select_columns, row)))
+                )
+
+        else:
+            ref_cache = LRUCache(max_size=1000)
+
+            batch_size = 10
+            batch = []
+            for row in raw_res:
+                call_dict = _ch_call_dict_to_call_schema_dict(
+                    dict(zip(select_columns, row))
+                )
+                batch.append(call_dict)
+
+                if len(batch) >= batch_size:
+                    hydrated_batch = self._hydrate_calls(
+                        batch, req.expand_columns, ref_cache
+                    )
+                    for call in hydrated_batch:
+                        yield tsi.CallSchema.model_validate(call)
+
+                    # *** Dynamic Batch Size ***
```

Comment:
there are edges where this will fail the 1000 limit. I think you should probably make a way to skip that 1000 limit in these cases so that we don't accidentally fail an entire request do to an invalid estimation here.

Reply:
will have to anyway to do the above comment (using `_refs_read_batch_within_project`)